app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -91,15 +162,16 @@ def save_inventory_single(type, parent, hash, child_keys = [], extra_keys = [],
     save_child_inventory(child, hash, child_keys)
   end
 
-  def save_inventory_with_findkey(association, hash, deletes, new_records, record_index)
+  def save_inventory_with_findkey(association, hash, deletes, new_records, record_index, id_only = false)
     # Find the record, and update if found, else create it
     found = record_index.fetch(hash)
     if found.nil?
-      found = association.build(hash.except(:id))
-      new_records << found
+      found = association.public_send(id_only ? :new : :build, hash.except(:id))
+      new_records << found if new_records
     else
+      found = association.find(found) if id_only
```

Comment:
is this doing a query for each updated record? That would cause quite a bloat and slowdown on a remote DB

Reply:
This is where the `if` comes into play when using this, so this is only called with the `#save_inventory_multi` code, and not with `#save_inventory_with_thin_association`

```ruby
# from #save_inventory_with_thin_association
found = save_inventory_with_findkey(model, h.except(*remove_keys).merge(assoc_attrs), deletes_index, nil, record_index, true)
```

Notice that `nil` is passed in for `new_records`, so this will never get called from `save_inventory_with_thin_association`.  However, this will get called when `save_inventory_multi` is called for anything within `save_inventory_with_thin_association`, but that was always the case.

Reply:
Hmm... ignore that comment... for some reason, it looked like it was pointing to a different line when viewing it in the "Files Changed" part...

Reply:
I see ```found = save_inventory_with_findkey(model, h.except(*remove_keys).merge(assoc_attrs), deletes_index, nil, record_index, true)``` so e.g. when doing second refresh of 50k images, this would do 50k queries to fetch them, right?

Reply:
So yes, with `found = association.find(found) if id_only`, I am not sure how to get around it without basically skipping callbacks and doing a straight SQL update:

```sql
UPDATE table
SET val_1 = 1, val_2 = 'two', val_3 = 'iii'
WHERE id = 1
```

Which may or may not be okay.  Again, the solving for Memory v.s. CPU makes this tough, and it might be something where using batches is the solution here as a comprimize.

Should definitely do some tests to see if "refresh N" is now slower than it was previously, since I was only really doing benchmarks on the first refresh.

Reply:
Right, so best to track the number of queries too, since on remote DB those can do around 5ms per query. In some models(e.g. Vm), we still have a business logic in hooks, so we need the old way of saving. I would like to slowly get rid of those (the batch SQL is not using hooks, those are moved to post refresh actions or as a connected custom graph node).